### ISSUE TITLE
Optimize random access

### ISSUE THREAD

**Issue by Ralith** [Ralith/hecs#177]
Fetching individual components currently requires hitting the hash table that maps component types to columns every time. For applications which do large amounts of random access for the same component types (e.g. rapier) this is redundant. We could optimize this case by precomputing an archetype -> column offset mapping for a given type up front, in a helper that borrows the `World` and hence will be disposed before any spawns might invalidate it. This would also allow us to perform dynamic borrow checking only once.

Implementation outline:
```rust
impl World {
    fn column<T: Component>(&self) -> Column<'_, T>;
    fn column_mut<T: Component>(&self) -> ColumnMut<'_, T>;
}

struct Column<'a, T> {
    entities: &'a [EntityMeta],
    archetypes: &'a [Archetype],
    archetype_column_offsets: Vec<Option<u32>>,
    _marker: PhantomData<T>,
}

impl<T> Column<'_, T> {
    fn get(&self, entity: Entity) -> Result<&T, ComponentError>;
}
```

**Comment by Ralith** [comment]
It might also be worth cacheing the column offset tables, rather than computing them on-demand. A `TypeIdMap<Vec<Option<u32>>>`, updated whenever a new archetype is added or an existing archetype is realloced, would do the trick.

**Comment by Ralith** [comment]
Implemented in #206.
